### PR TITLE
`Paywalls`: extracted configuration processing into a new `TemplateViewConfiguration`

### PR DESCRIPTION
**Pull request by NachoSoto**
### Changes:
- Created `TemplateViewConfiguration` to encapsulate all processed configuration for a paywall
- Created `TemplateViewConfiguration.PackageSetting` and `TemplateViewConfiguration.PackageConfiguration` to represent paywalls that support 1 or N packages
- Taken all common logic out of `Example1Template`
- Added new `PaywallViewMode` for upcoming "ramps" views

### CODE REVIEW

File: RevenueCatUI/Templates/Example1Template.swift [hunk 1]
```diff
@@ -4,101 +4,63 @@ import SwiftUI
...
+                                introEligibility: self.introEligibility)
+        .task(id: self.package) {
+            self.introEligibility = await self.introEligibilityChecker.eligibility(for: self.package)
         }
     }
 
-    private var package: Package? {
-        return (try? self.data.get())?.package
+    private var package: Package {
+        return self.configuration.packages.single.content
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Example1Template.swift:28]: Example of how templates access the desired package configuration

File: RevenueCatUI/Templates/Example1Template.swift [hunk 2]
```diff
@@ -4,101 +4,63 @@ import SwiftUI
...
     private var purchaseHandler: PurchaseHandler
     @Environment(\.dismiss)
     private var dismiss
 
-    init(data: Data, introEligibility: IntroEligibilityStatus?) {
-        self.data = data
+    init(configuration: TemplateViewConfiguration, introEligibility: IntroEligibilityStatus?) {
+        self.configuration = configuration
         self.introEligibility = introEligibility
+        self.localization = configuration.packages.single.localization
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/Example1Template.swift:48]: Example of how templates access the desired package configuration

File: RevenueCatUI/Templates/TemplateViewType.swift [hunk 3]
```diff
@@ -1,57 +1,63 @@
...
+        return Result {
+            TemplateViewConfiguration(
+                mode: mode,
+                packages: try .create(with: offering.availablePackages,
+                                      filter: self.config.packages,
+                                      localization: self.localizedConfiguration,
+                                      setting: self.template.packageSetting),
+                configuration: self.config,
+                colors: self.config.colors.multiScheme,
+                headerImageURL: self.headerImageURL
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/TemplateViewType.swift:50]: This is now all the processing for a paywall in one place: filtering packages, combining colors, and processing localization.

File: RevenueCatUI/Templates/TemplateViewType.swift [hunk 4]
```diff
@@ -1,57 +1,63 @@
...
-            } else {
-                Logger.warning("Couldn't find '\(type)'")
-                return nil
-            }
+    @ViewBuilder
+    private static func createView(template: PaywallTemplate,
+                                   configuration: TemplateViewConfiguration) -> some View {
+        switch template {
+        case .example1:
+            Example1Template(configuration)
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/TemplateViewType.swift:60]: Very simple `switch` now from `PaywallTemplate` to each view, with just 1 configuration value.

File: RevenueCatUI/Templates/TemplateViewType.swift [hunk 5]
```diff
@@ -1,57 +1,63 @@
...
+    init(_ configuration: TemplateViewConfiguration)
+
+}
+
+@available(iOS 16.0, macOS 13.0, tvOS 16.0, *)
+private extension PaywallTemplate {
+
+    var packageSetting: TemplateViewConfiguration.PackageSetting {
+        switch self {
+        case .example1: return .single
```
Comment (NachoSoto) [inline RevenueCatUI/Templates/TemplateViewType.swift:17]: Very simple `switch` from `PaywallTemplate` to `TemplateViewConfiguration.PackageSetting`.
When we add a new template, this makes it obvious that we need to specify this setting.

**Review by joshdholtz — APPROVED** [review]
One small typo in error but other than this this looks super clean 😊

File: RevenueCatUI/Data/TemplateError.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        ]
+    }
+
+    private var description: String {
+        switch self {
+        case .noPackages:
+            return "Attempted to display paywall with no packages."
+
+        case .emptyPackageList:
+            return "Paywall configuration container no packages."
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply RevenueCatUI/Data/TemplateError.swift:39]: Thanks!

**Pull request merged**